src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/AzureMetadataRequestor.cs

```diff
@@ -0,0 +1,105 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Net;
+    using System.Threading.Tasks;
+#if NETSTANDARD1_3
+    using System.Net.Http;
+#endif
+
+    internal class AzureMetadataRequestor : IAzureMetadataRequestor
+    {
+        /// <summary>
+        /// Azure Instance Metadata Service exists on a single non-routable IP on machines configured
+        /// by the Azure Resource Manager. See <a href="https://go.microsoft.com/fwlink/?linkid=864683">to learn more.</a>
+        /// </summary>
+        private static string baseImdsUrl = $"http://169.254.169.254/metadata/instance/compute";
+        private static string imdsApiVersion = $"api-version=2017-04-02"; // this version has the format=text capability
+        private static string imdsTextFormat = "format=text";
+
+        /// <summary>
+        /// Gets the value of a specific field from the imds link asynchronously, and returns it.
+        /// </summary>
+        /// <param name="fieldName">Specific imds field to retrieve a value for</param>
+        /// <returns>an array of field names available, or null</returns>
+        public async Task<string> GetAzureComputeMetadata(string fieldName)
+        {
+            string metadataRequestUrl = $"{baseImdsUrl}/{fieldName}?{imdsTextFormat}&{imdsApiVersion}";
+
+            string requestResult = await this.MakeAzureMetadataRequest(metadataRequestUrl);
+
+            return requestResult;
+        }
+
+        /// <summary>
+        /// Gets all the available fields from the imds link asynchronously, and returns them as an IEnumerable.
+        /// </summary>
+        /// <returns>an array of field names available, or null</returns>
+        public async Task<IEnumerable<string>> GetAzureInstanceMetadataComputeFields()
+        {
+            string allFieldsResponse = string.Empty;
+            string metadataRequestUrl = $"{baseImdsUrl}?{imdsTextFormat}&{imdsApiVersion}";
+
+            allFieldsResponse = await this.MakeAzureMetadataRequest(metadataRequestUrl);
+
+            string[] fields = allFieldsResponse?.Split(Environment.NewLine.ToCharArray(), StringSplitOptions.RemoveEmptyEntries);
+
+            if (fields == null || fields.Length <= 0)
+            {
+                CoreEventSource.Log.CannotObtainAzureInstanceMetadata();
+            }
+
+            return fields;
+        }
+
+        private async Task<string> MakeAzureMetadataRequest(string metadataRequestUrl)
+        {
+            string requestResult = string.Empty;
+
+            SdkInternalOperationsMonitor.Enter();
+            try
+            {
+#if NETSTANDARD1_3
+
+                using (var getFieldValueClient = new HttpClient())
+                {
+                    getFieldValueClient.DefaultRequestHeaders.Add("Metadata", "True");
+                    requestResult = await getFieldValueClient.GetStringAsync(metadataRequestUrl).ConfigureAwait(false);
```

Comment:
dont we need to ensure http response code is OK before reading response? (we had a bug in X-Component lookup similar to this, where we didn't check for OK before reading respone)

Reply:
I believe in this case we can disregard this, as the entire module that reads IMS data is done in a Task that is completely ignored in terms of return value. If the response is not OK, and there is some problem raised here (which would throw an exception perhaps) then the Task terminates and the data for IMS is not injected into the heartbeat payload. At most, we would want to get a log line here I suspect.